### DISCUSSION TITLE
Support for OIDC extensions: OIDC4IDA

### DISCUSSION THREAD

**Discussion by tnorimat** [keycloak/keycloak#21270] [Ideas]
Hello everyone,

I would like to propose supporting [OpenID Connect for Identity Assurance 1.0 (OIDC4IDA)](https://openid.net/specs/openid-connect-4-identity-assurance-1_0.html), one of extensions of OpenID Connect.

It can append identity assurance data on authenticated user to ID token and UserInfo endpoint.
When registering a user on IdP:
- How their identity was verified (e.g., Validation that physical evidence is genuine through inspection of its physical properties in person).
- Which framework and policy was applied (e.g., eIDAS, [UK digital identity and attributes trust framework](https://www.gov.uk/government/publications/uk-digital-identity-and-attributes-trust-framework-beta-version), UK Good Practice Guide [(GPG)](https://www.gov.uk/government/publications/identity-proofing-and-verification-of-an-individual))
- Which evidence was used (e.g., driver license, information from Mobile Network Operator).

This extension enables RP/Client app to check whether the authentication user can be creadible flexibly.

**Comment by tnorimat** [comment]
The prototype implementation for discussion is https://github.com/keycloak/keycloak/pull/21309 .

**Comment by thomasdarimont** [comment]
Part of adding support for the OpenID Connect 4 identity assurance is to add [additional claims about end-users](https://openid.net/specs/openid-connect-4-identity-assurance-1_0.html#name-additional-claims-about-end).

One step towards this goal is having the proper constants and (built-in) protocol mappers available to emit the proper token claims.
![image](https://github.com/keycloak/keycloak/assets/314690/7d39d0d9-9753-4bdc-8a9f-04cd2aa24165)
Most of the implementations could be implemented as attribute mappers and registered with the [OIDCLoginProtocolFactory](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/protocol/oidc/OIDCLoginProtocolFactory.java#L153) as a first step.

Will this be considered in the context of this effort or should this be covered in a different discussion?
I could work on this if this is not already covered.

**Comment by tnorimat** [comment]
@thomasdarimont Yes, if Keycloak itself provides such the claims, we need to add some mappers for them as you suggested.
However, [its prototype implementation](https://github.com/keycloak/keycloak/pull/21309) takes such the strategy for constructing claims that it asks for them to some entity like Claim Provider (both internal and external), where Keycloak itself needs not prepare such the additional claims about end-users.

**Comment by Bredstone** [comment]
Hello everyone! Following suggestions, I'm here to share with you my solution proposal for implementing OIDC4IDA support on Keycloak, using a custom service provider, and ensuring modularity. I've detailed it further in #30152.

The idea behind it was to rely on an [open-source library](https://github.com/authlete/authlete-java-common) provided by Authlete to handle the protocol's message processing. This library has official support for OIDC4IDA, [is listed as one of the reference implementations](https://bitbucket.org/openid/ekyc-ida/wiki/Implementations) and is even maintened by an organization that actively works on the protocol.

Another keypoint of my approach was to keep it modular and ensure it would work with several Keycloak instances regardless of version.

There are some improvements points to be made, for sure, like the support to user's consent and figuring out a way to store users' attributes, but it is currently a work in progress project. Maybe we could work together to improve it.

I am also looking forward to list this solution on the [official protocol implementations page](https://bitbucket.org/openid/ekyc-ida/wiki/Implementations) in order to bring visibility to the project and help people who might be benefited by this :)
